src/geom/util/GeometryFixer.cpp

```diff
@@ -0,0 +1,346 @@
+    bool isMixed = false;
+    for (std::size_t i = 0; i < p_geom->getNumGeometries(); i++) {
+        const LineString* line = p_geom->getGeometryN(i);
+        if (line->isEmpty())
+            continue;
+
+        std::unique_ptr<Geometry> fix = fixLineStringElement(line);
+        if (fix == nullptr)
+            continue;
+
+        if (fix->getGeometryTypeId() != GEOS_LINESTRING) {
+            isMixed = true;
+        }
+        fixed.emplace_back(fix.release());
+    }
+    if (fixed.size() == 1) {
+        std::unique_ptr<Geometry> g(fixed.at(0).release());
+        return g;
+    }
+    if (isMixed)
+        return factory->createGeometryCollection(std::move(fixed));
+    else
+        return factory->createMultiLineString(std::move(fixed));
+}
+
+/* private */
+std::unique_ptr<Geometry>
+GeometryFixer::fixPolygon(const Polygon* p_geom)
+{
+    std::unique_ptr<Geometry> fix = fixPolygonElement(p_geom);
+    if (fix == nullptr)
+        return factory->createPolygon();
+    else
+        return fix;
+}
+
+/* private */
+std::unique_ptr<Geometry>
+GeometryFixer::fixPolygonElement(const Polygon* p_geom)
+{
+    const LinearRing* shell = p_geom->getExteriorRing();
+    std::unique_ptr<Geometry> fixShell = fixRing(shell);
+    if (fixShell->isEmpty()) {
+        if (isKeepCollapsed) {
+            const CoordinateSequence* cs = shell->getCoordinatesRO();
+            std::unique_ptr<LineString> line(factory->createLineString(*cs));
+            return fixLineString(line.get());
+        }
+        //-- if not allowing collapses then return empty polygon
+        return nullptr;
+    }
+    // if no holes then done
+    if (p_geom->getNumInteriorRing() == 0) {
+        return fixShell;
+    }
+    std::unique_ptr<Geometry> fixedHoles = fixHoles(p_geom);
+    std::unique_ptr<Geometry> result = removeHoles(fixShell.get(), fixedHoles.get());
+    return result;
+}
+
+/* private */
+std::unique_ptr<Geometry>
+GeometryFixer::fixHoles(const Polygon* p_geom)
+{
+    std::vector<std::unique_ptr<Geometry>> holes;
+
+    for (std::size_t i = 0; i < p_geom->getNumInteriorRing(); i++) {
+        std::unique_ptr<Geometry> holeRep = fixRing(p_geom->getInteriorRingN(i));
+        // Do not preserve null/empty holes
+        if (holeRep != nullptr && !holeRep->isEmpty()) {
+            holes.emplace_back(holeRep.release());
+        }
+    }
+    if (holes.empty())
+        return nullptr;
+
+    if (holes.size() == 1) {
+        std::unique_ptr<Geometry> h(holes.at(0).release());
+        return h;
+    }
+    std::unique_ptr<GeometryCollection> holesGeom = (factory->createGeometryCollection(std::move(holes)));
+    return operation::overlayng::OverlayNGRobust::Union(holesGeom.get());
+}
+
+/* private */
+std::unique_ptr<Geometry>
+GeometryFixer::removeHoles(const Geometry* shell, const Geometry* holes)
+{
+    if (holes == nullptr || holes->isEmpty())
+        return shell->clone();
+    return operation::overlayng::OverlayNGRobust::Overlay(shell, holes,
```

Comment:
Why call a specific implementation instead of `shell->difference(holes)` ?

Reply:
Just following the JTS implementation... I guess this one has the fallbacks for sure (which the default does now too!)

Reply:
This is because in JTS OverlayNG is not (yet) the default.  In GEOS it's fine to use the `Geometry.difference()` method.